### PR TITLE
	apply scope check for all Role related commands

### PR DESCRIPTION
**Pull request by wisexp**
## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [ ] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by wisexp** [comment]
@darshanhs90 is added to the review. #Closed

**Comment by wisexp** [comment]
@markcowl is added to the review. #Closed

**Comment by wisexp** [comment]
@azuresdkci is added to the review. #Closed

File: src/ResourceManager/Resources/Commands.Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs [hunk 1]
```diff
@@ -237,6 +237,8 @@ public override void ExecuteCmdlet()
                 ExcludeAssignmentsForDeletedPrincipals = true
             };
 
+            AuthorizationClient.ValidateScope(options.Scope);
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs]: >AuthorizationClient.ValidateScope(options.Scope); [](start = 12, length = 49)

we can validate before creating the options right,if the scope is invalid,there isnt any use of the creation of options

File: src/ResourceManager/Resources/Commands.Resources/RoleAssignments/NewAzureRoleAssignmentCommand.cs [hunk 2]
```diff
@@ -167,6 +167,8 @@ public override void ExecuteCmdlet()
                 }
             };
 
+            AuthorizationClient.ValidateScope(parameters.Scope);
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/RoleAssignments/NewAzureRoleAssignmentCommand.cs]: similar to the other one

File: src/ResourceManager/Resources/Commands.Resources/RoleAssignments/RemoveAzureRoleAssignmentCommand.cs [hunk 3]
```diff
@@ -179,6 +179,8 @@ public override void ExecuteCmdlet()
                 IncludeClassicAdministrators = false
             };
 
+            AuthorizationClient.ValidateScope(options.Scope);
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/RoleAssignments/RemoveAzureRoleAssignmentCommand.cs]: >AuthorizationClient.ValidateScope(options.Scope); [](start = 12, length = 49)

similar to the other one

File: src/ResourceManager/Resources/Commands.Resources/RoleDefinitions/GetAzureRoleDefinitionCommand.cs [hunk 4]
```diff
@@ -65,6 +65,8 @@ public override void ExecuteCmdlet()
                 RoleDefinitionName = Name,
             };
 
+            AuthorizationClient.ValidateScope(options.Scope);
+
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/RoleDefinitions/GetAzureRoleDefinitionCommand.cs]: similar to the other one

File: src/ResourceManager/Resources/Commands.Resources/RoleDefinitions/RemoveAzureRoleDefinitionCommand.cs [hunk 5]
```diff
@@ -75,6 +75,8 @@ public override void ExecuteCmdlet()
                 }
             };
 
+            AuthorizationClient.ValidateScope(options.Scope);
+
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources/RoleDefinitions/RemoveAzureRoleDefinitionCommand.cs]: similar to the other one

File: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs [hunk 6]
```diff
@@ -62,5 +62,14 @@ public void RDGetScenario()
...
             ResourcesController.NewInstance.RunPsTest("Test-RDGet");
         }
+
+        [Fact]
+        [Trait(Category.AcceptanceType, Category.CheckIn)]
+        public void RdValidateInputParameters() 
+        {
+            var instance = ResourcesController.NewInstance;
+            instance.RunPsTest("Test-RdValidateInputParameters Get-AzureRmRoleDefinition");
+            instance.RunPsTest("Test-RdValidateInputParameters Remove-AzureRmRoleDefinition");
```
Comment (darshanhs90) [inline src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleDefinitionTests.cs]: >"Test-RdValidateInputParameters Remove-AzureRmRoleDefinition"); [](start = 31, length = 63)

no test for new-azurermroledefinition?

**Comment by darshanhs90** [comment]
> }

even new-azurermroledefinition has a list of assignable scopes,shouldnt we validate those aswell?

---
Refers to: src/ResourceManager/Resources/Commands.Resources/RoleDefinitions/GetAzureRoleDefinitionCommand.cs:73 in 2369a24. [](commit_id = 2369a24610f89c737987182a9fb47b15d6f2fe82, deletion_comment = False)

File: src/ResourceManager/Resources/Commands.Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs (hunk 1 above)
Comment (wisexp) [inline src/ResourceManager/Resources/Commands.Resources/RoleAssignments/GetAzureRoleAssignmentCommand.cs]: options.Scope could be different than the input scope.

---
In reply to: [125784541](https://github.com/Azure/azure-powershell/pull/4228#discussion_r125784541) [](ancestors = 125784541)

**Comment by darshanhs90** [comment]
:shipit:

**Comment by wisexp** [comment]
>     Assert-Throws { Get-AzureRmRoleAssignment -Scope $badScope } $badScopeException

this is no longer correct, and this scenario is already covered by Test-RaValidateInputParameters

---
Refers to: src/ResourceManager/Resources/Commands.Resources.Test/ScenarioTests/RoleAssignmentTests.ps1:64 in 99f4a67. [](commit_id = 99f4a6744ac838cd1db7a15fe8571e5e40187264, deletion_comment = True)

**Comment by darshanhs90** [comment]
@shahabhijeet could you take a look into this PR?

File: .gitignore [hunk 7]
```diff
@@ -203,4 +203,9 @@ FakesAssemblies/
 /tools/*.dll
 *.GhostDoc.xml
 pingme.txt
-groupMapping*.json
\ No newline at end of file
+groupMapping*.json
```
Comment (markcowl) [inline .gitignore]: these additions to the gitignore are incorrect.  Please revert changes to this file unless you have  a file you would liek to add.

Comment (darshanhs90) [inline .gitignore]: @markcowl reverted .gitignore

**Comment by darshanhs90** [comment]
@markcowl reverted gitignore, as per the comments

**Comment by markcowl** [comment]
@wisexp Could you please summarize these changes in the changelog? You can do this in a separate PR.

on demand run here:http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/1753/

**Pull request merged**